homeassistant/components/airthings_ble/sensor.py

```diff
@@ -107,9 +115,65 @@
 }
 
 
+def migrate(hass: HomeAssistant, address: str, sensor_name: str):
+    """Migrate entities to new unique ids (with BLE Address)."""
+
+    ent_reg = entity_async_get(hass)
+
+    device_registry = device_async_get(hass)
+    entity_registry = entity_async_get(hass)
+
+    device = device_registry.async_get_device(identifiers={(DOMAIN, address)})
+
+    if not device:
+        return
+
+    entities = async_entries_for_device(
+        entity_registry,
+        device_id=device.id,
+        include_disabled_entities=True,
+    )
+
+    filtered_entities = (
+        entity for entity in entities if entity.unique_id.endswith(sensor_name)
+    )
+
+    unique_ids: dict[str, str] = {}
+
+    for entity in filtered_entities:
+        if entity.unique_id.startswith(address):
+            unique_ids["v3"] = entity.entity_id
+        elif "(" in entity.unique_id:
+            unique_ids["v2"] = entity.entity_id
+        else:
+            unique_ids["v1"] = entity.entity_id
+
+    def _migrate_unique_id(entity_id: str, new_unique_id: str):
+        ent_reg.async_update_entity(entity_id=entity_id, new_unique_id=new_unique_id)
+        _LOGGER.info("Migrated entity '%s' to unique id '%s'", entity_id, new_unique_id)
+
+    if unique_ids.get("v3"):
+        # Already has the newest unique id format
+        return
+
+    new_unique_id = f"{address}_{sensor_name}"
+
+    # Try to migrate the pre 2023.9.0 unique id first
+    if entity_id := unique_ids.get("v1"):
+        _migrate_unique_id(
+            entity_id=entity_id,
+            new_unique_id=new_unique_id,
+        )
+
```

Comment:
Since there is so much branching here, can you add a test for migrating the ID from both formats?